src/main/kotlin/com/bardsoftware/backend/fleet/rmarkdown/Subscriber.kt

```diff
@@ -22,35 +22,100 @@ import com.google.cloud.pubsub.v1.Subscriber
 import com.google.pubsub.v1.PubsubMessage
 import com.google.pubsub.v1.SubscriptionName
 import com.xenomachina.argparser.ArgParser
+import org.apache.commons.io.FileUtils.readFileToString
+import java.io.ByteArrayInputStream
+import java.io.File
+import java.io.FileOutputStream
+import java.nio.charset.Charset
+import java.nio.file.Path
+import java.nio.file.Paths
 import java.util.concurrent.CompletableFuture
+import java.util.zip.ZipEntry
+import java.util.zip.ZipInputStream
 
 class SubscriberArgs(parser: ArgParser) {
     val subscriberName by parser.storing(
             "--sub",
             help = "subscription topic name")
+
+    val tasksDir by parser.storing(
+            "--tasks-dir",
+            help = "root of tasks content"
+    )
 }
 
 private val PROJECT_ID = ServiceOptions.getDefaultProjectId()
 
-internal class MessageReceiverExample(private val callback: (message: String, md5sum: String) -> Unit) : MessageReceiver {
+internal class TaskReceiver(private val tasksDir: Path,
+                            private val callback: (message: String, md5sum: String) -> Unit
+) : MessageReceiver {
     private val dockerProcessor = DockerProcessor()
+    private val BUFFER_SIZE = 4096
+
 
     override fun receiveMessage(message: PubsubMessage, consumer: AckReplyConsumer) {
         processMessage(message)
         consumer.ack()
     }
 
     fun processMessage(message: PubsubMessage) {
-        val messageContent = message.data.toStringUtf8()
-        val md5sum = this.dockerProcessor.getMd5Sum(messageContent)
+        val request = CompilerFleet.CompilerFleetRequest.parseFrom(message.data)
+
+        val taskId = request.taskId
+        val destination = this.tasksDir.resolve(taskId).resolve("files")
+        val zipByteInput = ByteArrayInputStream(request.zipBytes.toByteArray())
+        val zipStream = ZipInputStream(zipByteInput)
+        var entry: ZipEntry? = zipStream.nextEntry
+        val buffer = ByteArray(this.BUFFER_SIZE)
+
+        while (entry != null) {
+            val filename = entry.name
+            val newFile = destination.resolve(filename).toFile()
+
+            File(newFile.parent).mkdirs()
+            val fos = FileOutputStream(newFile)
+
+            var len: Int = zipStream.read(buffer)
+            while (len > 0) {
+                fos.write(buffer, 0, len)
+                len = zipStream.read(buffer)
+            }
+
+            fos.close()
+            entry = zipStream.nextEntry
+        }
 
-        this.callback(messageContent, md5sum)
+        val rootFileName = request.rootFileName
+        val pathToRoot = destination.resolve(rootFileName).toFile()
+        if (!pathToRoot.exists()) {
+            throw RootFileNotFoundException()
+        }
+
+        val content = readFileToString(pathToRoot, Charset.defaultCharset())
```

Comment:
I suggest changing docker command instead, so that it could cat the given file and calculate md5. This will move us closer to running the real compiler.

Reply:
Changed, but not quite, what I wanted.
Is it good to put all file content like this?

I've tried to copy to container file and get md5 sum via shell command, but docker client seems can copy only directories. I looked up through their tests, how they are using copyToContainer command. 
File should be either directory or archive. It's true also for input streams.

Reply:
As I understand, [this issue](https://github.com/spotify/docker-client/issues/943) still isn't closed